File: composeApp/src/androidMain/kotlin/org/iremia/iremia/ui/journal/CalendarMath.kt
package org.iremia.iremia.ui.journal

import kotlinx.datetime.DateTimeUnit
import kotlinx.datetime.DayOfWeek
import kotlinx.datetime.LocalDate
import kotlinx.datetime.daysUntil
import kotlinx.datetime.isoDayNumber
import kotlinx.datetime.plus
import java.text.DateFormatSymbols
import java.util.Calendar

// =============================================================================
// Pure date helpers for the Journal calendar.
//
// Kept free of Compose so the week/month math stays trivially testable and can be
// reused by both the collapsed (week) and expanded (month) calendar views.
// Weeks are Monday-first, matching the design and the German locale.
// =============================================================================

/** Number of days in [month] (1-based) of [year]. */
fun daysInMonth(year: Int, month: Int): Int {
    val first = LocalDate(year, month, 1)
    return first.daysUntil(first.plus(1, DateTimeUnit.MONTH))
}

/**
 * The seven dates of a rolling week starting at [start] (inclusive).
 *
 * Used for the collapsed strip so "today" stays the leftmost cell and the next six
 * days follow, instead of snapping to a fixed Monday→Sunday week.
 */
fun rollingWeek(start: LocalDate): List<LocalDate> =
    (0..6).map { start.plus(it, DateTimeUnit.DAY) }

/** Localized short weekday label for [date] (device locale), e.g. "Sun". */
fun shortWeekdayLabel(date: LocalDate): String {
    val calendarDay = when (date.dayOfWeek) {
        DayOfWeek.MONDAY -> Calendar.MONDAY
        DayOfWeek.TUESDAY -> Calendar.TUESDAY
        DayOfWeek.WEDNESDAY -> Calendar.WEDNESDAY
        DayOfWeek.THURSDAY -> Calendar.THURSDAY
        DayOfWeek.FRIDAY -> Calendar.FRIDAY
        DayOfWeek.SATURDAY -> Calendar.SATURDAY
        DayOfWeek.SUNDAY -> Calendar.SUNDAY
        else -> Calendar.MONDAY
    }
    return DateFormatSymbols.getInstance().shortWeekdays[calendarDay].trimEnd('.')
}

/**
 * Monday-first calendar grid for [month] (1-based) of [year].
 *
 * Leading/trailing cells are `null` so the 1st lands under its weekday column and
 * the final week row is always complete (multiple of 7 cells).
 */
fun monthGrid(year: Int, month: Int): List<LocalDate?> {
    val leading = LocalDate(year, month, 1).dayOfWeek.isoDayNumber - 1
    val days = daysInMonth(year, month)
    val cellCount = ((leading + days + 6) / 7) * 7 // round up to whole weeks
    return (0 until cellCount).map { index ->
        val dayNumber = index - leading + 1
        if (dayNumber in 1..days) LocalDate(year, month, dayNumber) else null
    }
}

/** A specific month of a specific year. */
data class YearMonth(val year: Int, val month: Int)

/**
 * A window of consecutive months centered on [date]'s month, from [monthsBefore]
 * months in the past to [monthsAfter] months in the future. Used to feed the
 * scrollable multi-month (expanded) calendar.
 */
fun monthsAround(date: LocalDate, monthsBefore: Int, monthsAfter: Int): List<YearMonth> {
    val firstOfMonth = LocalDate(date.year, date.monthNumber, 1)
    return (-monthsBefore..monthsAfter).map { offset ->
        val shifted = firstOfMonth.plus(offset, DateTimeUnit.MONTH)
        YearMonth(shifted.year, shifted.monthNumber)
    }
}

/** Localized Monday → Sunday short weekday labels (device locale), e.g. ["Mo", …]. */
fun weekdayLabels(): List<String> {
    val short = DateFormatSymbols.getInstance().shortWeekdays
    val mondayFirst = listOf(
        Calendar.MONDAY, Calendar.TUESDAY, Calendar.WEDNESDAY, Calendar.THURSDAY,
        Calendar.FRIDAY, Calendar.SATURDAY, Calendar.SUNDAY,
    )
    // Trim the trailing dot some locales add (e.g. German "Mo.").
    return mondayFirst.map { short[it].trimEnd('.') }
}

/** Localized full month name for [month] (1-based), e.g. "April" (device locale). */
fun monthName(month: Int): String =
    DateFormatSymbols.getInstance().months[month - 1]

File: iosApp/iosApp/Views/JournalViews/Prototype/CalendarMath.swift
import Foundation

// =============================================================================
// Pure date helpers for the Journal calendar.
// 1:1 translation of CalendarMath.kt.
// Weeks are Monday-first, matching the design and the German locale.
// =============================================================================

/// Number of days in a given month (1-based) of a given year.
func daysInMonth(year: Int, month: Int) -> Int {
    var comps = DateComponents()
    comps.year = year
    comps.month = month
    let cal = Calendar(identifier: .iso8601)
    guard let date = cal.date(from: comps) else { return 30 }
    return cal.range(of: .day, in: .month, for: date)?.count ?? 30
}

/// The seven dates of a rolling week starting at `start` (inclusive).
/// Used for the collapsed strip so "today" stays the leftmost cell.
func rollingWeek(start: Date) -> [Date] {
    let cal = Calendar(identifier: .iso8601)
    return (0..<7).compactMap { cal.date(byAdding: .day, value: $0, to: start) }
}

/// Localized short weekday label for a date, e.g. "Mon".
func shortWeekdayLabel(date: Date) -> String {
    let fmt = DateFormatter()
    fmt.locale = Locale.autoupdatingCurrent
    fmt.dateFormat = "E"
    let label = fmt.string(from: date)
    // Trim trailing dot some locales add (e.g. German "Mo.")
    return label.hasSuffix(".") ? String(label.dropLast()) : label
}

/// A specific month of a specific year.
struct YearMonth: Hashable, Identifiable {
    let year: Int
    let month: Int
    var id: Int { year * 100 + month }
}

/// Monday-first calendar grid for a month. Leading/trailing nil cells
/// so the 1st lands under its weekday column.
func monthGrid(year: Int, month: Int) -> [Date?] {
    let cal = Calendar(identifier: .iso8601)
    var comps = DateComponents()
    comps.year = year
    comps.month = month
    comps.day = 1
    guard let firstDay = cal.date(from: comps) else { return [] }

    // ISO weekday: Monday = 1, Sunday = 7
    let weekday = cal.component(.weekday, from: firstDay)
    // Convert to Monday-first index (0 = Monday)
    let mondayBased = (weekday + 5) % 7
    let days = daysInMonth(year: year, month: month)
    let cellCount = ((mondayBased + days + 6) / 7) * 7

    return (0..<cellCount).map { index in
        let dayNumber = index - mondayBased + 1
        if dayNumber >= 1 && dayNumber <= days {
            var dc = DateComponents()
            dc.year = year
            dc.month = month
            dc.day = dayNumber
            return cal.date(from: dc)
        }
        return nil
    }
}

/// A window of consecutive months centered on a date.
func monthsAround(date: Date, monthsBefore: Int, monthsAfter: Int) -> [YearMonth] {
    let cal = Calendar(identifier: .iso8601)
    let year = cal.component(.year, from: date)
    let month = cal.component(.month, from: date)

    return (-monthsBefore...monthsAfter).map { offset in
        var totalMonth = (year * 12 + month - 1) + offset
        let y = totalMonth / 12
        let m = totalMonth % 12 + 1
        return YearMonth(year: y, month: m)
    }
}

/// Localized Monday → Sunday short weekday labels, e.g. ["Mon", …, "Sun"].
func weekdayLabels() -> [String] {
    let cal = Calendar(identifier: .iso8601)
    let fmt = DateFormatter()
    fmt.locale = Locale.autoupdatingCurrent
    // Build dates for a known Monday→Sunday week
    var comps = DateComponents()
    comps.year = 2024
    comps.month = 1
    comps.day = 1 // Monday
    let monday = cal.date(from: comps)!

    return (0..<7).map { offset in
        let day = cal.date(byAdding: .day, value: offset, to: monday)!
        return shortWeekdayLabel(date: day)
    }
}

/// Localized full month name for a 1-based month, e.g. "April".
func monthName(month: Int) -> String {
    let fmt = DateFormatter()
    fmt.locale = Locale.autoupdatingCurrent
    return fmt.monthSymbols[month - 1]
}

// MARK: - Date comparison helpers

/// Extract day-of-month from a Date.
func dayOfMonth(_ date: Date) -> Int {
    Calendar(identifier: .iso8601).component(.day, from: date)
}

/// Check if two dates are the same calendar day.
func isSameDay(_ a: Date, _ b: Date) -> Bool {
    Calendar(identifier: .iso8601).isDate(a, inSameDayAs: b)
}

/// Get today's date at start of day.
func startOfToday() -> Date {
    Calendar(identifier: .iso8601).startOfDay(for: Date())
}
